config/kernel-blkdev.m4

```diff
@@ -539,6 +540,20 @@ AC_DEFUN([ZFS_AC_KERNEL_SRC_BLKDEV_ISSUE_SECURE_ERASE], [
 		    sector, nr_sects, GFP_KERNEL);
 	])
 
+	ZFS_LINUX_TEST_SRC([blkdev_issue_discard_async_flags], [
+		#include <linux/blkdev.h>
+	],[
+		struct block_device *bdev = NULL;
+		sector_t sector = 0;
+		sector_t nr_sects = 0;
+		unsigned long flags = 0;
+		struct bio *biop = NULL;
+		int error __attribute__ ((unused));
+
+		error = __blkdev_issue_discard(bdev,
+		    sector, nr_sects, GFP_KERNEL, flags, &biop);
```

Comment:
This interface changed again in the 5.18 kernel you'll need to check for both.  Commit torvalds/linux@44abff2c0b970ae3d310b97617525dc01f248d7c.

5.18+ API:
```
 int __blkdev_issue_discard(struct block_device *bdev, sector_t sector,
     sector_t nr_sects, gfp_t gfp_mask, struct bio **biop)
```

Reply:
@behlendorf I think we can manage without explicitly checking for 5.18+ API change. The commit that introduces `blkdev_issue_secure_erase`, i.e. https://github.com/torvalds/linux/commit/44abff2c0b970ae3d310b97617525dc01f248d7c, removes the `flags` argument from `__blkdev_issue_discard` and `blkdev_issue_discard`. The same commit also removes the `BLKDEV_DISCARD_SECURE` flag.

Checking for the presence of secure erase API, i.e. `blkdev_issue_secure_erase` confirms whether to use `__blkdev_issue_discard` and `blkdev_issue_discard` from prior to v5.18, where `flags` argument is used or later where flags argument is removed from both `__blkdev_issue_discard` and `blkdev_issue_discard`. So that's what we are doing here. If `HAVE_BLKDEV_ISSUE_SECURE_ERASE` is defined, we assume `__blkdev_issue_discard` and `blkdev_issue_discard` are also present without `flags` argument. This also helps to reduce #ifdefs ion the code and keep it cleaner. 

This stays aligned with previous code, where we were not checking for 5.18+ API change for `blkdev_issue_discard`, whether it takes flags argument or not. We relied on presence of `blkdev_issue_secure_erase` to confirm whether to call `blkdev_issue_discard` with flags argument previously as well.

Reply:
I see, yes that should be workable.